Validate LLM API key before registering it

timtmok opened the pull request:
Address #6616

Adds some methods to be able to register one model at a time. It performs a request on the model and registers it with the language API if it succeeded.

<!-- Thank you for submitting a pull request.
If this is your first pull request you can find information about
contributing here:
  * https://github.com/posit-dev/positron/blob/main/CONTRIBUTING.md

We recommend synchronizing your branch with the latest changes in the
main branch by either pulling or rebasing.
-->

<!--
  Describe briefly what problem this pull request resolves, or what
  new feature it introduces. Include screenshots of any new or altered
  UI. Link to any GitHub issues but avoid "magic" keywords that will
  automatically close the issue. If there are any details about your
  approach that are unintuitive or you want to draw attention to, please
  describe them here.
-->


### Release Notes

<!--
  Optionally, replace `N/A` with text to be included in the next release notes.
  The `N/A` bullets are ignored. If you refer to one or more Positron issues,
  these issues are used to collect information about the feature or bugfix, such
  as the relevant language pack as determined by Github labels of type `lang: `.
  The note will automatically be tagged with the language.

  These notes are typically filled by the Positron team. If you are an external
  contributor, you may ignore this section.
-->

#### New Features

- Validates LLM API key before adding it to Positron Assistant #6616

#### Bug Fixes

- N/A


### QA Notes
There will be an error toast notification if the model was not added.


https://github.com/user-attachments/assets/17712e81-75ce-4be2-969f-9150bedef386



<!--
  Add additional information for QA on how to validate the change,
  paying special attention to the level of risk, adjacent areas that
  could be affected by the change, and any important contextual
  information not present in the linked issues.
-->

timtmok committed:
Tests AI connection before adding it

Sends a test hello message to the LLM

timtmok committed:
Allow registering one model at a time

Displays error message if the model could not be added

timtmok committed:
Refactoring

jmcphers commented on extensions/positron-assistant/src/models.ts:
```diff
@@ -97,6 +105,25 @@ abstract class AILanguageModel implements positron.ai.LanguageModelChatProvider
+			// if the model responds, the config works
+			return true;
+		} catch (error) {
+			return false;
```
Can we read this error and show it to the user so they have some idea why the connection failed? Could be a network issue, bad API key, etc.

timtmok replied:
`ai-sdk` returns an APICallError that has some details from Anthropic about what happened. I haven't checked other AI's but this seems to be as granular as it gets with the API call. I could return a more detail from the response like the status code that might be useful for an admin but likely not useful for a user.

![image](https://github.com/user-attachments/assets/ec19b3bc-7e6b-43df-89c2-0424ff40012a)

jmcphers commented on extensions/positron-assistant/src/models.ts:
```diff
@@ -97,6 +105,25 @@ abstract class AILanguageModel implements positron.ai.LanguageModelChatProvider
+			// send a test message to the model
+			await ai.generateText({
+				model: this.model,
+				prompt: 'Hello!',
```
This can return a pretty wordy response with some models and since each token takes some time to generate, I think we should use a prompt that is likely to be faster. Maybe like this?

<img width="234" alt="image" src="https://github.com/user-attachments/assets/2c84773c-f9b0-463f-a34d-83910776c998" />

jmcphers commented on extensions/positron-assistant/src/models.ts:
```diff
@@ -97,6 +105,25 @@ abstract class AILanguageModel implements positron.ai.LanguageModelChatProvider
+
+		try {
+			// send a test message to the model
+			await ai.generateText({
```
This can take an unbounded amount of time during which there's no user feedback -- should we show a progress indicator?

timtmok replied:
That's a good idea. This could probably be improved further with the dialog remaining open if an error occurred (as well as showing what went wrong).

![image](https://github.com/user-attachments/assets/fc3f27b4-1e5d-4ace-8b33-e4f95d91874b)

jmcphers reviewed (dismissed):
Confirmed that the model doesn't get saved if I use a bad API key, LGTM modulo some suggestions for improving the UX

timtmok committed:
Improve error handling

Add progress bar to dialog
Show error details on failure
Change prompt so AI provides a short response
